### ISSUE TITLE
Add `--cheatsheet` option to `duckpan new` command. 

### ISSUE THREAD

**Issue by moollaza** [duckduckgo/p5-app-duckpan#268]
Now that we've made it so easy to create a CheatSheet it seems like a `duckpan new --cheatsheet` command could help to make things a bit simpler. It should ideally generate the correct file, with the correct filename, in the correct location (/json). It could also pre-fill some of the info, or at least provide a skeleton of the document.

So far most developers seem to be doing OK, but this might be a useful tool if anyone wants to build it?

**Comment by srvsh** [comment]
@moollaza, if someone is not working on this already, may I try and do this one?

**Comment by moollaza** [comment]
@zachthompson did you have plans to work on this?

@srvsh AFAIKm no one else is working on this :+1:

**Comment by zachthompson** [comment]
Please do, @srvsh

**Comment by srvsh** [comment]
Thanks! I'll start work on this.

**Comment by zachthompson** [comment]
Thank you @srvsh!  We have wanted to be able to add general template processing to duckpan for a while.  If you want to open a PR early, we would be happy to follow your progress and provide feedback.

**Comment by srvsh** [comment]
@zachthompson thanks, I'll open a PR early on for feedback. Is there already some documentation about how general template processing is planned to be done? It'll be useful to ensure that my changes would fit in with the overall plan.

**Comment by zachthompson** [comment]
@srvsh I haven't looked at `App::DuckPAN::Cmd::New` in too much detail recently.  However, one maybe ideal goal would be to have the ability to add arbitrary template files and just have them work.  The `new` command will clearly need to be much more dynamic.

Give it a shot and let's see what happens!

**Comment by moollaza** [comment]
@srvsh there isn't any formal documentation on how the current templating works, but it can be reasoned from the code.

We have in the Spice and Goodies repos a "template" dir with some blank files that contain sections which are replaced.

The DuckPAN code takes user input, determines which repo you're currently in (Spice/Goodies), grabs the template file and then creates new files from the templates, placing them in the correct place, and fills them appropriate with the user input.

We *could* move the templates out of the repos, but then they need to be placed somewhere else that DuckPAN can access. Likely it's own /share/ directory?

**Comment by srvsh** [comment]
Thanks @zachthompson, @moollaza . Here's my proposal for this

* We can move the template information from DuckPAN.pm to a new YAML file.  There will be one such file for the different instant answer repositories.  Tentatively, I'm thinking of having it in `templates/templates.yml`. Here's the proposed file structure:
```
# file: zeroclickinfo-goodies/template/templates.yml
---
# templates for goodies

default:
    # template file. These are present in the 'template' subdirectory
    lib/DDG/Goodie/Example.pm:
        # output file. The string within {{..}} will be replaced
        # with a computed value. In this case, it'll be the 
        # instant answer's name in CamelCase
        lib/DDG/Goodie/{{ia_name_cc}}.pm
    t/Example.t:
        t/{{ia_name_cc}}.t

# a cheatsheet
cheatsheet:
    share/cheat_sheet.json:
        share/goodie/cheat_sheets/json/{{ia_name_lc}}.json
...
```
* Create a new `App::DuckPAN::Templates` class in the `p5-app-duckpan` repository. This would handle reading the `templates.yml` file to understand which templates can be use for the instant answer. This class would also handle creation of actual files from the templates. There would also have to be some logic to maintain backward compatibility for when the YAML file was absent. Apart from this, there would have to be methods to get a list of all template files since `App::DuckPAN::Restart` would need this to watch the containing directories for changes.

* Add a new optional `--template <template_name>` argument to the `new` command. That way, we know which template the user wants to use and create files accordingly. If this argument is not provided, we use the `default` template.

Thanks!

**Comment by moollaza** [comment]
@srvsh this seems like a good start to me.

I'd like to here @zachthompson thoughts though

**Comment by zachthompson** [comment]
@srvsh I like your proposal.

With respect to backward compatibility, I'm leaning toward a warning that their repo isn't up to date.  If someone is starting a new instant answer they should probably be working off the latest master with the template config.  We could even prompt to automatically update it for them.

Don't worry about the `restart` behavior since you really can't generate new IAs while `query` or `server` is running.

One other thought would be to have `required` and `optional` sections within each template type.  The user could then be prompted to configure optional files if they exist.  For example, goodies can optionally have javascript and css files like [spices](https://github.com/duckduckgo/zeroclickinfo-spice/tree/master/template/share/spice/example) (even though these templates aren't currently in the repo!) and spices don't require css but do javascript.

**Comment by srvsh** [comment]
@zachthompson thanks for your inputs! I'm now clear about the restarts. I'll incorporate your suggestions about the `required` and `optional` sections and send out a pull request in a few hours.

**Comment by srvsh** [comment]
@moollaza @zachthompson I apologise for the delay. I have raised a pull request #279 to support arbitrary templates along with optional files. The design is more or less the same as proposed except for some additions to support optional files. Kindly take a look and let me know what you think.

I have updated my forks of the goodie and spice repositories to include the YAML files and demonstrate the features:
Goodie: https://github.com/srvsh/zeroclickinfo-goodies/tree/generic_templates
Spice: https://github.com/srvsh/zeroclickinfo-spice/tree/generic_templates_throwaway

The Spice repo has optional files just for demonstration purposes, so I'll change it later :)

While working on this, I saw that the `ia_name_separated` variable in `App::DuckPAN::Cmd::New` does not have names separated by spaces (for example, it'll be `HelloWorld`, but I thought it would be `Hello World`). Should that be fixed?

Thanks!

**Comment by moollaza** [comment]
>While working on this, I saw that the ia_name_separated variable in App::DuckPAN::Cmd::New does not have names separated by spaces (for example, it'll be HelloWorld, but I thought it would be Hello World). Should that be fixed?

@srvsh good catch. You can fix that if you like, but very soon, we won't be using that stuff (the Perl metadata) anymore, as we now have an Instant Answer Page for each IA which stores all that metadata. 

Actually, @zachthompson I guess we should remove the metadata stuff now while we're working on the templates?

**Comment by zachthompson** [comment]
yep, @moollaza.  @srvsh I have branches for [goodies](https://github.com/duckduckgo/zeroclickinfo-goodies/compare/zach/templates) and [spices](https://github.com/duckduckgo/zeroclickinfo-spice/compare/zach/templates) that update Example.pm to how we want it.  If you want to use those as a base that would be great.  We also need the [share templates](https://github.com/duckduckgo/zeroclickinfo-spice/tree/master/template/share/spice/example) for goodies but the javascript will be slightly different.

**Comment by srvsh** [comment]
@moollaza thanks, I needed the name to be in separate words to generate the `name` field in the cheatsheet JSON files. I'll go ahead and fix it up.

@zachthompson Sure, I'll rebase my commits onto your branches and send out an updated pull request along with the tests in a couple of days. Adding the share templates for goodies should be straightforward, so I can add it whenever the javascript is finalised. Or if you think I can update the javascript for goodies, then I'll be happy to take it up as well.

Thanks!

**Comment by zachthompson** [comment]
@srvsh those template changes have been merged :D

**Comment by srvsh** [comment]
@zachthompson that's great :) I've now rebased my changes onto the latest master and updated the pull request. I added test cases and found there was an issue using `each` for iterating over hashes, so that's another commit that's been pushed.

I'll need to write code to notify the user in case their IA repositories are not recent enough to work with this templating method. I should be able to get to it on Monday.

Thanks!

**Comment by srvsh** [comment]
@zachthompson I think I'm done with this and opened pull requests in the relevant packages. Do let me know what you think :)

Here are the new commands added to the duckpan app:
```
$ duckpan new --list-templates # lists out the available templates for the current instant answer type
Available templates are:
cheatsheet, default

$duckpan new # uses the 'default' template
Please enter a name for your Instant Answer:  : Test Goodie
Do you want to create files for "css" [y/n]?  : y
Do you want to create files for "javascript" [y/n]?  : y
Created file: lib/DDG/Goodie/TestGoodie.pm
Created file: share/goodie/test_goodie/test_goodie.js
Created file: t/TestGoodie.t
Created file: share/goodie/test_goodie/test_goodie.css
Successfully created Goodie: TestGoodie

$duckpan new --template cheatsheet
Please enter a name for your Instant Answer:  : DuckDuckGo   
Created file: share/goodie/cheat_sheets/json/duck_duck_go.json
Successfully created Goodie: DuckDuckGo

$ duckpan new --template bad_template # invalid template
[FATAL]  Template 'bad_template' does not exist for this IA Type: Goodie. Please use one of: cheatsheet, default

# If run in an older goodie repository with no templates.yml, prompt the user to update it
$ duckpan new
[FATAL]  No templates exist for this IA Type: Goodie. Maybe you need to pull the latest version of this repository?

```
I have written test cases for the templating logic and have tried using the generated files by writing some simple Goodies (default and cheatsheet). I still have to test Spice, but hopefully there won't be issues since there's almost no change there. The restarter also restarts the server on modification of any code or cheatsheet json.

Thanks!

**Comment by zachthompson** [comment]
Fantastic @srvsh I will be looking at this asap!

**Comment by srvsh** [comment]
@moollaza @zachthompson Can this issue be closed now that PR #279 has been merged?

To summarise, there are two ways of creating a cheatsheet now:
1. `duckpan new -c`
2. `duckpan new --template cheatsheet`

**Comment by zachthompson** [comment]
I believe so :+1:
